doc/man3/OPENSSL_s390xcap.pod

```diff
@@ -0,0 +1,224 @@
+
+The environment variable is a semicolon-separated list of tokens which is
+processed from left to right (whitespace is ignored):
+
+ OPENSSL_s390xcap="<tok1>;<tok2>;..."
+
+There are three types of tokens:
+
+=over 4
+
+=item <string>
+
+The name of a processor generation. A bit in the environment variable's
+mask is set to one if and only if the specified processor generation
+implements the corresponding instruction set extension. Possible values
+are z900, z990, z9, z10, z196, zEC12, z13 and z14.
+
+=item <string>:<mask>:<mask>
+
+The name of an instruction followed by two 64-bit masks. The part of the
+environment variable's mask corresponding to the specified instruction is
+set to the specified 128-bit mask. Possible values are kimd, klmd, km, kmc,
+kmac, kmctr, kmo, kmf, prno and kma.
+
+=item stfle:<mask>:<mask>:<mask>
+
+Store-facility-list-extended (stfle) followed by three 64-bit masks. The
+part of the environment variable's mask corresponding to the stfle
+instruction is set to the specified 192-bit mask.
+
+=back
+
+The 64-bit masks are specified in hexadecimal notation. The 0x prefix is
+optional. Prefix a mask with a tilde (~) to denote a bitwise NOT operation.
+
+The following bits are significant ...
+
+=over 4
+
+=item stfle:
+
+ #38 store-clock-fast facility
+ #46 message-security assist
+ :
+ #50 message-security assist extension 4
+ #51 message-security assist extension 3
+ :
+ #45 message-security assist extension 8
+ #56 vector enhancements facility 1
+ #57 vector packed decimal facility
+ #62 vector facility
+
+=back
+
+=over 4
+
+=item kimd:
+
+ #26 KIMD-SHAKE-256
+ #27 KIMD-SHAKE-128
+ #28 KIMD-SHA3-512
+ #29 KIMD-SHA3-384
+ #30 KIMD-SHA3-256
+ #31 KIMD-SHA3-224
+ #60 KIMD-SHA-512
+ #61 KIMD-SHA-256
+ #62 KIMD-SHA-1
+ :
+ #62 KIMD-GHASH
+
+=back
+
+=over 4
+
+=item klmd:
+
+ #26 KLMD-SHAKE-256
+ #27 KLMD-SHAKE-128
+ #28 KLMD-SHA3-512
+ #29 KLMD-SHA3-384
+ #30 KLMD-SHA3-256
+ #31 KLMD-SHA3-224
+ :
+
+=back
+
+=over 4
+
+=item km:
+
+ #11 KM-XTS-AES-256
+ #13 KM-XTS-AES-128
+ #43 KM-AES-256
+ #44 KM-AES-192
+ #45 KM-AES-128
+ :
+
+=back
+
+=over 4
+
+=item kmc:
+
+ #43 KMC-AES-256
+ #44 KMC-AES-192
+ #45 KMC-AES-128
+ :
+
+=back
+
+=over 4
+
+=item kmac:
+
+ #43 KMAC-AES-256
+ #44 KMAC-AES-192
+ #45 KMAC-AES-128
+ :
+
+=back
+
+=over 4
+
+=item kmctr:
+
+ :
+
+=back
+
+=over 4
+
+=item kmo:
+
+ #43 KMO-AES-256
+ #44 KMO-AES-192
+ #45 KMO-AES-128
+ :
+
+=back
+
+=over 4
+
+=item kmf:
+
+ #43 KMF-AES-256
+ #44 KMF-AES-192
+ #45 KMF-AES-128
+ :
+
+=back
+
+=over 4
+
+=item prno:
+
+ :
+ #13 PRNO-TRNG
+
+=back
+
+=over 4
+
+=item kma:
+
+ #43 KMA-GCM-AES-256
+ #44 KMA-GCM-AES-192
+ #45 KMA-GCM-AES-128
+ :
+
+=back
+
+=head1 EXAMPLES
+
+=over 4
+
+=item OPENSSL_s390xcap="z196"
+
+Disables all instruction set extensions which the z196 processor does not implement.
+
+=item OPENSSL_s390xcap="stfle:~0:~0:~0x4000000000000000"
+
+Disables the vector facility.
+
+=item OPENSSL_s390xcap="km:~0x2800:~0;kimd:~0xc000000:~0"
```

Comment:
I wonder if following would be usable/appreciated. `~0x100` *alone* or `~0x100:` would mean that second word is zeroed. While `:~0x100` would mean first work is not modified.

Reply:
> [... ] ~0x100: would mean that second word is zeroed. While :~0x100 would mean first work is not modified.

Did you mean an omitted mask value is interpreted as ~0 ("dont modify") ? .. the quote suggests otherwise: omitting second word means 0 ("zeroise"), ommitting first word means ~0 ("dont modify"), so i dont get the logic here.

As it is now, you have to specify both words i.e., keep the ~0 if you dont want to modify detected capabilities.

Reply:
First of all I want to emphasize that remark started with "I wonder". In other words it's rather a question than actual suggestion.

> Did you mean an omitted mask value is interpreted as ~0 ("dont modify") ?

For first word, yes, but opposite for second. So that `:~0x100` would be equivalent to `~0:~0x100`, but `~0x100` would be equivalent to `~0x100:0`. I recognize that it might appear "quirky". It might be more appropriate to arrange that omitted value is always equivalent to ~0...

Reply:
It appears "quirky" because i dont get the rationale behind it. I mean, why would one want to make disabling all higher function codes (second word) the default if one decides to disable a lower (first word) function code ? E.g. if one disables sha3 fcs for kimd, why should disabling ghash fc be the default?

As for "omitted value means ~0": this makes sense to me. Its a trade of between not requiring the user to specify ~0 and adding complexity to parsing code.

Reply:
> E.g. if one disables sha3 fcs for kimd, why should disabling ghash fc be the default?

Just in case, rationale was based on assumption that new stuff is always denoted by more significant bits. So that if user wanted to suppress something new, chances are that [s]he wants to suppress even everything newer. In this case assuming 0 for second word would be kind of intuitive. This assumption turns to be wrong in this case. It's all right :-)